receiver/awsfirehosereceiver/internal/unmarshaler/cwmetricstream/unmarshaler.go

```diff
@@ -39,53 +47,58 @@ func NewUnmarshaler(logger *zap.Logger) *Unmarshaler {
 // Skips invalid cWMetrics received in the record and
-func (u Unmarshaler) Unmarshal(records [][]byte) (pmetric.Metrics, error) {
-	md := pmetric.NewMetrics()
-	builders := make(map[resourceAttributes]*resourceMetricsBuilder)
-	for recordIndex, record := range records {
-		// Multiple metrics in each record separated by newline character
-		for datumIndex, datum := range bytes.Split(record, []byte(recordDelimiter)) {
-			if len(datum) > 0 {
-				var metric cWMetric
-				err := json.Unmarshal(datum, &metric)
-				if err != nil {
-					u.logger.Error(
-						"Unable to unmarshal input",
-						zap.Error(err),
-						zap.Int("datum_index", datumIndex),
-						zap.Int("record_index", recordIndex),
-					)
-					continue
-				}
-				if !u.isValid(metric) {
-					u.logger.Error(
-						"Invalid metric",
-						zap.Int("datum_index", datumIndex),
-						zap.Int("record_index", recordIndex),
-					)
-					continue
-				}
-				attrs := resourceAttributes{
-					metricStreamName: metric.MetricStreamName,
-					namespace:        metric.Namespace,
-					accountID:        metric.AccountID,
-					region:           metric.Region,
-				}
-				mb, ok := builders[attrs]
-				if !ok {
-					mb = newResourceMetricsBuilder(md, attrs)
-					builders[attrs] = mb
-				}
-				mb.AddMetric(metric)
+func (u Unmarshaler) UnmarshalMetrics(record []byte) (pmetric.Metrics, error) {
+	metrics := pmetric.NewMetrics()
+
+	// Multiple metrics in each record separated by newline character
+	decoder := json.NewDecoder(bytes.NewReader(record))
+	for datumIndex := 0; ; datumIndex++ {
+		var cwMetric cWMetric
+		if err := decoder.Decode(&cwMetric); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
 			}
+			u.logger.Error(
+				"Unable to unmarshal input",
+				zap.Error(err),
+				zap.Int("datum_index", datumIndex),
+			)
+			continue
 		}
+		if !u.isValid(cwMetric) {
+			u.logger.Error(
+				"Invalid metric",
+				zap.Int("datum_index", datumIndex),
+			)
+			continue
+		}
+
+		rm := metrics.ResourceMetrics().AppendEmpty()
+		setResourceAttributes(cwMetric, rm.Resource())
+
+		metric := rm.ScopeMetrics().AppendEmpty().Metrics().AppendEmpty()
+		metric.SetName(cwMetric.MetricName)
+		metric.SetUnit(cwMetric.Unit)
+
+		dp := metric.SetEmptySummary().DataPoints().AppendEmpty()
+		dp.SetTimestamp(pcommon.NewTimestampFromTime(time.UnixMilli(cwMetric.Timestamp)))
+		setDataPointAttributes(cwMetric, dp)
+		dp.SetCount(uint64(cwMetric.Value.Count))
+		dp.SetSum(cwMetric.Value.Sum)
+		minQ := dp.QuantileValues().AppendEmpty()
+		minQ.SetQuantile(0)
+		minQ.SetValue(cwMetric.Value.Min)
+		maxQ := dp.QuantileValues().AppendEmpty()
+		maxQ.SetQuantile(1)
+		maxQ.SetValue(cwMetric.Value.Max)
 	}
 
-	if len(builders) == 0 {
-		return pmetric.NewMetrics(), errInvalidRecords
+	if metrics.MetricCount() == 0 {
+		return metrics, errInvalidRecords
 	}
 
-	return md, nil
+	metrics = expmetrics.Merge(pmetric.NewMetrics(), metrics)
```

Comment:
Again would like a benchmark.  This feels like a lot of duplicated work for inputs of any appreciable size.

Reply:
Done, and like in cwlog I've reverted to grouping by the known CloudWatch attributes, rather than using `expmetrics.Merge` here. Note that there's still merging at the consumer level, since at that level we don't know anything about the data -- so it has to merge in a more generic and expensive way.

Reply:
Addressed in my other merged PR, resolving.